csharp/lang/security/cryptography/X509-subject-name-validation.yaml

```diff
@@ -0,0 +1,66 @@
+rules:
+- id: X509-subject-name-validation
+  severity: WARNING
+  languages:
+  - csharp
+  metadata:
+    cwe: 'CWE-295: Improper Certificate Validation'
+    owasp:
+    - A02:2017 - Broken Authentication
+    - A07:2021 - Identification and Authentication Failures
+    references:
+    - https://docs.microsoft.com/en-us/dotnet/api/system.identitymodel.tokens.issuernameregistry?view=netframework-4.8
+    category: security
+    technology:
+    - .net
+    confidence: MEDIUM
+  message: 'Validating certificates based on subject name is '
+  patterns:
+  - pattern-inside: |
+      using System.IdentityModel.Tokens;
+      ...
+  - pattern-either:
+    - patterns:
+      - pattern-either:
+        - pattern-inside: |
+            X509SecurityToken $TOK = $RHS;
+            ...
+        - pattern-inside: |
+            $T $M(..., X509SecurityToken $TOK, ...) {
+                ...
+            }
+      - metavariable-pattern:
+          metavariable: $RHS
+          pattern-either:
+          - pattern: $T as X509SecurityToken
+          - pattern: new X509SecurityToken(...)
+    - patterns:
+      - pattern-either:
+        - pattern-inside: |
+            X509Certificate2 $CERT = new X509Certificate2(...);
```

Comment:
Modern C# makes heavy use of inferred typing with `var` - I'm not sure C# fully supports typed metavars yet but this will miss inferred typing.

Reply:
Doesn't look like it does yet. I 
* Changed the LHS types on those 3 `pattern-inside` lines and tried using 
  * `metavariable-pattern` with a `pattern-either` of `var` or `X509Certificate2`
  * `metavariable-regex` on `(var|X509Certificate2)`
* And added another loop that uses `var` instead of `X509Certificate2` in the test code

but the `var` loop wasn't caught in any of those cases